scripts/new_upstream_snapshot.py

```diff
@@ -41,10 +41,134 @@ class CliError(Exception):
     pass
 
 
+class VersionInfo:
+    def __init__(
+        self,
+        major: int,
+        minor: int,
+        hotfix: Optional[int] = None,
+        *,
+        debian: Optional[int] = None,
+        ubuntu: Optional[int] = None,
+        series: Optional[str] = None,
+        series_revision: Optional[int] = None,
+        pre_revision: Optional[int] = None,
+        pre_commit: Optional[str] = None,
+    ):
+        if any((series, series_revision)) and any((pre_revision, pre_commit)):
+            raise RuntimeError("Cannot contain both series and pre_version")
+        self.major = major
+        self.minor = minor
+        self.hotfix = hotfix
+        self.debian = debian
+        self.ubuntu = ubuntu
+        self.series = series
+        self.series_revision = series_revision
+        self.pre_revision = pre_revision
+        self.pre_commit = pre_commit
+
+    @classmethod
+    def from_string(cls, version):
+        # Something like 23.1.1-0ubuntu1~22.04.1
+        # or 23.1~1g111f1a6e-0ubuntu1
+        pattern = (
+            r"(?P<major>\d+)"
+            r"\."
+            r"(?P<minor>\d+)"
+            r"((\.(?P<hotfix>\d+))|(~(?P<pre_revision>\d+)g(?P<pre_commit>\S{8})))?"  # noqa: E501
+            r"(-(?P<debian>\d+))"
+            r"(ubuntu(?P<ubuntu>\d+))"
```

Comment:
Per maintainer's handbook, this regex may not match for non-cloud-init cases as downstream could be `ubuntuX.Y but for cloud-init we don't care here I don't think as we'll generally by syncing our releases down from devel which generally should be 0ubuntu1 or 0ubuntu0 with a supplemental ~XX.YY.# series suffix. So, maybe this is a "don't care" for this tool.